### ISSUE TITLE
Audible "clicks" on note start

### ISSUE THREAD

**Issue by thecoreyburton** [adplug/adplug#54]
In certain IMF tracks (I've specifically noticed it in a lot of Bobby Prince's work), there are audible clicks at the start of certain notes. 

As a fair warning, I've found that once I knew the clicks were present, I couldn't forget, and it does hinder the listening experience somewhat. If this sort of thing is likely to irritate you then proceed with caution: [wonderin.zip](https://github.com/adplug/adplug/files/1443221/wonderin.zip)

Above is an example file with the problem present. It's the menu song from Wolfenstein 3D (renamed from .imf to .wlf) "WONDERIN.WLF". Using my own build from the latest source, I played back this file using WoodyOPL at 49716hz, 16-bit and had the stereo option selected. I used the disk-writer plugin to export the first five seconds of playback to .wav

There are several instances of the "clicks" throughout this short clip, but for example's sake I'll focus on one. If you playback the clip, at roughly 2.85 seconds there are three bass notes (before the melody repeats). On the first of these bass notes, right at the start, one such "click" is present. If you amplify the clip, the problem becomes visually apparent, with a large jump being present when the clip starts to play the next note. I'm not sure what the technical name for this issue is, but it looks very much like two clips hastily cut and joined together improperly and the "click" is from the mismatch in audio data.

I'm not entirely sure this is an adplug issue. It's likely an issue with the instrument, composition or even potentially the OPL emulator itself (I'm a bit out of my element here), but even so, I am curious as to what the cause is and whether (even if it's not an adplug issue) there can be a solution implemented in adplug. One such thought was some form of "anti-aliasing" on the adjoining edges to blend them together better (by the minimum number of samples required to make the click inaudible). If this were the case, so as not to compromise on accuracy, I'd suggest it being controlled by a checkbox in the UI, defaulting to off - but I know it's likely this is a feature that has to be implemented in the OPL emulator and not in adplug.

Additionally, it's worth noting that the problem is not exclusive to WoodyOPL (it's present in Nuked OPL3 as well, at the very least). It's also audible on the "surround" option of WoodyOPL, although the left and right channels don't necessarily exhibit the same behavior, with some channels having clicks when the other one doesn't and vice versa.

If this issue isn't something that can be tackled here then you can by all means close this.

**Comment by Malvineous** [comment]
This issue has bothered me for many years but alas it's not an AdPlug issue.  I don't understand the OPL emulator code too well, but as far as I'm aware, it uses lookup tables to generate some of the waveforms.  I suspect that in certain situations, a note is played when these lookup tables return the peak of the waveform, which means that note will go from silence to loud in one sample.  This sudden change in amplitude looks briefly like a square wave, which results in an audible click.

Over the years I have thought about ways around it but I'm not sure what the best solution is.  In AdPlug, the only thing we could do would be to apply a low-pass filter to smooth out the wave, however I think that would just make it a more muffled click instead of removing it entirely.

I think the only viable solution is to change the OPL emulator so that instead of playing a new note at 100% volume right away, each note is faded in over the space of a few dozen samples.  If done correctly this would delay the onset of the note by a small enough value that it wouldn't be musically noticeable, but it should remove the click.

Unfortunately I don't know enough about OPL emulators to really know where to begin with this.

Maybe you could raise the issue with the OPL emulator developers and see if they have any ideas?  I don't remember the original OPL chips having these clicks so I'm not sure what they did to avoid the issue.

**Comment by thecoreyburton** [comment]
I was afraid that would be the case. I'll see what I can do in terms of contacting the developers, thought I must say it feels good knowing I'm not the only one who has noticed this. 

Having implemented the OPL emulators into adplug, do you know the best way or place to contact both the person(s) behind WoodyOPL or Nuked OPL3?

**Comment by Malvineous** [comment]
For WoodyOPL you could try the [DOSBox forum](http://www.vogons.org), and for NukedOPL3 the author commented in #35 so you might be able to get in touch there.  I will also write @nukeykt here as that might send the author a notification too.

Given Nuked has a strong focus on emulation accuracy, I think that would probably be a good place to start.

**Comment by thecoreyburton** [comment]
I'll give time for that notification to do it's work and if nothing comes of it I'll try other ways of contacting him. I've also started a thread over on the DOSBox forums, so if you or anyone else reading this want to follow the discussion, here's the URL: http://www.vogons.org/viewtopic.php?f=32&t=56789

**Comment by nukeykt** [comment]
These clicks are present on real OPL2/3 as well. Here's sample from real OPL2 chip:
[opl2wolf1.zip](https://github.com/adplug/adplug/files/1448630/opl2wolf1.zip)

**Comment by Malvineous** [comment]
@nukeykt Very interesting, thanks for the audio sample!

What are your thoughts about ramping up the volume (over a hundred samples or so) for each channel upon note-on?  Is this practical to do in the emulator, and do you think it would remove the click without affecting anything else?

Is it the sort of thing you would consider having in the emulator, or do you favour a more faithful emulator, clicks and all?

**Comment by nukeykt** [comment]
OPL chip resets phase to 0 on each key on command. I think it is main reason of clicks.
I prefer to leave it as is because it is real hardware behavior. Possibly low pass filter will help to smooth it a bit.
Now my goal is to write cycle-accurate Yamaha FM emulators. So don't expect any "fix" on my code.

**Comment by Malvineous** [comment]
I've done some experimenting and if I modify NukedOPL to fade each note in over a few samples - even fading it in over a lot of samples so the song sounds obviously different - then it doesn't actually stop the clicking.  I now think the clicking is actually caused during the note *off* phase instead, which I think is much harder to handle.

**Comment by thecoreyburton** [comment]
"OPL chip resets phase to 0 on each key on command."
"I now think the clicking is actually caused during the note off phase instead,"

Just to further confirm both of these statements on a waveform level, it looks like the notes always start where they should; but when the next note plays, the last is cut off. This is concurrent with both of the above statements.

Perhaps a solution might be to perform a check, and extend the notes by x amount of samples (until the phase is back to 0) when it's toggled off. I don't know how difficult such an implementation would be or what kind of side effects this would have, however.

**Comment by Malvineous** [comment]
I was thinking the same thing, and it would be the ideal solution as ending the wave at the zero point would remove any click.

I think you'd have to investigate the various text-to-speech algorithms that use the OPL first though, and check how they make their sounds, in case they rely on this (although to be fair if the delay is only a few samples, I doubt it would affect anything in this regard.)

Unfortunately I'm not sure how practical it would be as it's possible to turn a note off and then immediately play another note on the same channel, so you'd have to somehow track that new note event and act on it a few dozen samples later, once the zero point has been reached.  It looks like Nuked has some kind of delayed event facility but I am not sure exactly what its purpose is, and I imagine it wouldn't be possible to use it for something like this.

**Comment by PalMusicFan** [comment]
Hello! It seems that I found a viable solution. 

After consulting an audio software designer, I was told that a new sound should not immediately terminate the old one. Instead, the old sound sould response to a KEY OFF event, and have its natual release. So, we need at least one more OPL unit in order to make it possible for multiple OPL units work in turn for each channel. 

I made two tests. 

I modified a player that plays a kind of OPL2 music (*.RIX) under DOS, made it use the extra channels provided by OPL3 as another OPL unit, together with the "first group" of channels, works in turn. It really elimites most of the "clicks"! 

I also modifed AdPlug's surround OPL. I made the right OPL as the other OPL. The two OPLs of L and R work in turn, and most of the "clicks" also disappeared! Yes, it is something like "ping pong" stereo, but I could use the foobar2000 DSP of "downmix to mono". I have tested the menu song from Wolfenstein 3D, pretty ideal. 

I say it elimites most of the "clicks", not all, because in a track I tested there were still some "clicks" that probably casued by a too long release of certain voice. Maybe three or more OPL units work in turn could solve this problem.
